### DISCUSSION TITLE
issue: Controller does not register correctly when disabled is true

### DISCUSSION THREAD

**Discussion by reinertisa** [react-hook-form/react-hook-form#8314] [Q&A] [answered]
### Version Number

7.30.0

### Codesandbox/Expo snack

https://codesandbox.io/s/jolly-jennings-7y1ndq?file=/src/index.js

### Steps to reproduce

1. Go to https://codesandbox.io/s/jolly-jennings-7y1ndq?file=/src/index.js
2. Click on 'Submit' button
3. Go to console

### Expected behaviour

When I click submit button first, I want to see **undefined** but instead returns "apple". When you click second and more times you will see undefined values in the console. Why does not return undefined at first render? 
<img width="278" alt="Screen Shot 2022-05-05 at 7 50 46 PM" src="https://user-images.githubusercontent.com/71358643/167059052-1dd8b7eb-d4ae-47b4-958b-7e9642a3c5fc.png">

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment, marked as answer]
inspect your onChange and what's gets sent to hook form, I would suggested sending `null` instead of `undefined`

```
<DownShift onChange={() => { // insepdct here field.onChange }} disabled={true} />
```

**Reply by reinertisa** [reply]
If it is disable, when I click Submit button, it does not affect onChange. I updated codesandbox. Can you check it out, please? https://codesandbox.io/s/jolly-jennings-7y1ndq?file=/src/DownShift.js

**Reply by bluebill1049** [reply]
yes, because react hook form has no context that input is disabled, then once you start changing the input send `undefined`, i would recommend you look into this example : https://codesandbox.io/s/usecontroller-own-state-wncet2?file=/src/App.tsx

**Reply by reinertisa** [reply]
@bluebill1049 I did not figure it out using onChange. Instead I figured it out in different way. Now it works but do you think this is right way? If it is not, please let me know. Updated codesandbox: https://codesandbox.io/s/jolly-jennings-7y1ndq?file=/src/index.js:570-605
I used  this codeblock. 
`<input {...getInputProps({ ...register(name, { disabled }) })} />`

**Reply by reinertisa** [reply]
When I looked at the your documentation about Controller API. So I think my solution does not look okay?
_Do not register input again. This component is made to take care of the registration process._
```
<Controller
  name="test"
  render={({ field }) => {
    // return <input {...field} {...register('test')} />; ❌ double up the registration
    return <input {...field} />; // ✅
  }}
/>
```

**Reply by reinertisa** [reply]
@bluebill1049 I fixed the problem and sharing updated codesandbox here. disabled works well. Still there are a lot of edge cases but if people improve it, they can focus on stateReducer function. As you mentioned before, I invoke onChange using useEffect depends on disabled value. updated codesandbox: https://codesandbox.io/s/jolly-jennings-7y1ndq?file=/src/index.js

**Comment by leapful** [comment]
@reinertisa , @bluebill1049 : 

In this use case, you can simplify the handling by using React.forwardRef() for your custom component DownShift and pass the "ref" prop to target input element.

Live demo: https://codesandbox.io/s/boring-microservice-m863e4?file=/src/DownShift.js

**Reply by reinertisa** [reply]
Thank you so much for your help. I fixed the problem @leapful . Check my solution, please. https://codesandbox.io/s/jolly-jennings-7y1ndq?file=/src/index.js

**Reply by bluebill1049** [reply]
thank you @leapful and glad you find out a solution @reinertisa (plus the sharing)

**Comment by reinertisa** [comment]
I have follow up question @leapful and @bluebill1049 . I updated @leapful solution . https://codesandbox.io/s/zen-surf-69tqfz?file=/src/index.js . I added checkbox to see value in input. When I made it disable, default value is gone. How to keep default value in input when it is disable and When it is enable, I entered a value in input but it throws type error. How can I figure it out? Thank you for your help.
